src/modes/ModeEstimation.jl

```diff
@@ -440,3 +442,190 @@ function _optimize(
+    DynamicPPL.link!(f.vi, spl)
+  else
+    DynamicPPL.invlink!(f.vi, spl)
+  end
+
+  return nothing
+end
+
+
+function transform2constrained(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  spl = DynamicPPL.SampleFromPrior()
+
+  linked = DynamicPPL.islinked(vi, spl)
+  
+  !linked && DynamicPPL.link!(vi, spl)
+  vi[spl] = par
+  DynamicPPL.invlink!(vi,spl)
+  tpar = vi[spl]
+  
+  linked && DynamicPPL.link!(vi,spl)
+
+  return tpar
+end
+
+function transform2constrained!(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  par .= transform2constrained(par, vi)
+  return nothing
+end
+
+function transform2unconstrained(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  spl = DynamicPPL.SampleFromPrior()
+
+  linked = DynamicPPL.islinked(vi, spl)
+  linked && DynamicPPL.invlink!(vi, spl)
+  vi[spl] = par
+  DynamicPPL.link!(vi, spl)
+  tpar = vi[spl]
+  !linked && DynamicPPL.invlink!(vi, spl)
+
+  return tpar
+end
+
+function transform2unconstrained!(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  par .= transform2unconstrained(par, vi)
+  return nothing
+end
+
+
+function _orderpar(par::NamedTuple, vi::DynamicPPL.VarInfo)
+  tmp_idx = indexin(collect(keys(vi.metadata)), collect(keys(par)))
+  tmp = collect(par)[tmp_idx]
+  return tmp, tmp_idx
+end
+
+
+function transform2unconstrained(par::NamedTuple, vi::DynamicPPL.VarInfo; order::Bool=true, array::Bool=true)
+  par_sor, par_idx = _orderpar(par, vi)
+
+  transform2unconstrained!(par_sor, vi)
+
+  if order
+    if array
+      return par_sor
+    else
+      return (; zip(keys(vi.metadata), par_sor)...)
+    end
+  else
+    if array
+      return par_sor[tmp_idx]
+    else
+      return (; zip(keys(par),par_sor[par_idx])...)
+    end
+  end
+end
+
+function transform2constrained(par::NamedTuple, vi::DynamicPPL.VarInfo; order::Bool=true, array::Bool=true)
+  par_sor, par_idx = _orderpar(par, vi)
+
+  transform2constrained!(par_sor, vi)
+
+  if order
+    if array
+      return par_sor
+    else
+      return (; zip(keys(vi.metadata), par_sor)...)
+    end
+  else
+    if array
+      return par_sor[tmp_idx]
+    else
+      return (; zip(keys(par),par_sor[par_idx])...)
+    end
+  end
+end
+
+
+function transform2constrained(res::Optim.MultivariateOptimizationResults, vi::DynamicPPL.VarInfo)
+  tres = deepcopy(res)
+  if !any(isnan.(tres.minimizer))
+    tres.minimizer = transform2constrained(tres.minimizer, vi)
+  else
+      @warn "Could not transform optimisation results due to NaNs in ':minimizer'."
+  end
+
+  if !any(isnan.(tres.initial_x))
+    tres.initial_x = transform2constrained(tres.initial_x, vi)
+  else
+    @warn "Could not transform initial values due to NaNs in ':initial_x'."
+  end
+
+  return tres
+end
+
+
+
+function instantiate_optimisation_problem(model::DynamicPPL.Model, ::MAP , ::unconstrained)
+  obj = OptimLogDensity(model, OptimizationContext(DynamicPPL.DefaultContext()))
+
+  transform!(obj)
+
+  init(init_vals) = transform2unconstrained(init_vals, obj.vi)
```

Comment:
IMO the initialization should be kept separate from defining the cost function and the transformation.

Reply:
It doesn't so much initialise but provides the user with transformation function (with the correct `vi`) together with the objective to transform the initial value into the correct space of the objective